choose-index: update the docs for supporting more patterns for mv index (#16855)
pingcap/docs-cn: choose-index.md

| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t6, index:idx(a, cast(`j` as signed array), b)  | range:[1 1,1 1], keep order:false, stats:pseudo                                                                                                                   |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                                                                   |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test2.t6.a, 1), or(eq(test2.t6.b, 1), eq(test2.t6.b, 2)), or(json_memberof(cast(1, json BINARY), test2.t6.j), json_memberof(cast(2, json BINARY), test2.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                                                                    |

> +-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Reviewer: 这里可以描述下还有更好的选择是什么（解释自己为啥不充分）

Author: 加一句“由于目前实现上的限制，`(b = 1 or b = 2)` 没有参与构成 IndexMerge，而是成为了一个单独的 Selection 算子”，然后加一个人工展开之后能充分利用的例子，这样呢？

Reviewer: 可以可以

Author: 改好了，再看看吧~

    | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo        |
    | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo        |
    | ├─IndexRangeScan_8(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[4,4], keep order:false, stats:pseudo        |
    | └─Selection_10(Probe)         | 0.00    | cop[tikv] |                                                                             | eq(json_length(json_extract(test.t4.j, "$.a")), 2) |
    |   └─TableRowIDScan_9          | 0.00    | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                     |
    +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+----------------------------------------------------+
    ```

> - 用于构成整体 IndexMerge 计划的每个子条件，需要分别符合用于连接的 `OR` 或 `AND` 的语义。具体如下：

Reviewer 2: ```suggestion
- 用于构成整体 IndexMerge 计划的每个子条件，需要分别符合用于连接的 `OR` 或 `AND` 的语义。具体如下：

```

        |   └─TableRowIDScan_7          | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                                                                                         |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
        ```

如果表达式包含嵌套的 `OR` 或 `AND`，或者表达式需要进行展开等变形之后才能直接一一对应各索引列，TiDB 可能无法使用 IndexMerge 或不能充分利用所有过滤条件。建议针对具体场景预先进行测试验证。以下是部分场景示例：

```sql
CREATE TABLE t5 (a INT, j JSON, b INT, k JSON, INDEX idx(a, (CAST(j AS SIGNED ARRAY))), INDEX idx2(b, (CAST(k as SIGNED ARRAY))));
CREATE TABLE t6 (a INT, j JSON, b INT, k JSON, INDEX idx(a, (CAST(j AS SIGNED ARRAY)), b), INDEX idx2(a, (CAST(k as SIGNED ARRAY)), b));
```

> 当 `AND` 嵌套在 `OR` 中，且嵌套的 `AND` 所连接的条件与多列索引的各个列能够一一对应时，TiDB 通常能够充分利用过滤条件。例如：

Reviewer 3: ```suggestion
当 `AND` 嵌套在由 `OR` 连接的条件中，且嵌套的 `AND` 所连接的子条件与多列索引的各个列能够一一对应时，TiDB 通常能够充分利用过滤条件。例如：
```

| id                            | estRows | task      | access object                                      | operator info                                   |
+-------------------------------+---------+-----------+----------------------------------------------------+-------------------------------------------------+
| IndexMerge_8                  | 0.20    | root      |                                                    | type: union                                     |
| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t5, index:idx(a, cast(`j` as signed array))  | range:[1 1,1 1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t5, index:idx2(b, cast(`k` as signed array)) | range:[2 2,2 2], keep order:false, stats:pseudo |
| └─TableRowIDScan_7(Probe)     | 0.20    | cop[tikv] | table:t5                                           | keep order:false, stats:pseudo                  |
+-------------------------------+---------+-----------+----------------------------------------------------+-------------------------------------------------+
```

> 当单个 `OR` 嵌套在 `AND` 中，且需要展开表达式才能直接对应索引列的时候，TiDB 通常能够充分利用过滤条件。例如：

Reviewer 3: ```suggestion
当单个 `OR` 嵌套在由 `AND` 连接的条件中，且 `OR` 所连接的子条件通过展开表达式可以直接对应索引列时，TiDB 通常能够充分利用过滤条件。例如：
```

| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t6, index:idx(a, cast(`j` as signed array), b)  | range:[1 1,1 1], keep order:false, stats:pseudo                                                                         |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                         |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test2.t6.a, 1), or(json_memberof(cast(1, json BINARY), test2.t6.j), json_memberof(cast(2, json BINARY), test2.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                          |
+-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
```

> 当多个 `OR` 嵌套在 `AND` 中，且需要展开表达式才能直接对应索引列的时候，TiDB 可能无法充分利用过滤条件。例如：

Reviewer 3: ```suggestion
当多个 `OR` 嵌套在由 `AND` 连接的条件中，且 `OR` 所连接的子条件需要展开表达式才能直接对应索引列的时候，TiDB 可能无法充分利用过滤条件。例如：
```

6 rows in set, 1 warning (0.00 sec)
```

对于多个 `json_member_of`、`json_contains` 或 `json_overlaps` 由 `OR`/`AND` 连接而成的条件，需要满足一些条件才能使用 IndexMerge 访问：

> 1. 对于由 `OR` 连接而成的条件，其每个子条件都需要分别能够使用 IndexMerge 访问；

Author: > 其每个子条件都需要分别都有 MVI 的覆盖；

这个意思？

> 对于多个 `json_member_of`、`json_contains` 或 `json_overlaps` 由 `OR`/`AND` 连接而成的条件，需要满足一些条件才能使用 IndexMerge 访问：
> 1. 对于由 `OR` 连接而成的条件，其每个子条件都需要分别能够使用 IndexMerge 访问；
> 2. 对于由 `AND` 连接而成的条件，其中一部分子条件需要分别能够使用 IndexMerge 访问，最终使用 IndexMerge 访问时也只能利用这一部分条件；
> 3. 用于构成整体 IndexMerge 计划的每个子条件需要分别符合用于连接的 `OR`/`AND` 的语义。具体来说，`json_contains` 由 `AND` 连接是符合语义的；`json_overlaps` 由 `OR` 连接是符合语义的；`json_member_of` 由 `OR`/`AND` 连接都是符合语义的；包含多个值的 `json_contains` 由 `OR` 连接或包含多个值的 `json_overlaps` 由 `AND` 连接是不符合语义的，而如果它们只包含一个值则是符合语义的。
> 例如：

Reviewer 2: ```suggestion
对于多个 `json_member_of`、`json_contains` 或 `json_overlaps` 由 `OR`/`AND` 连接而成的条件，需要满足以下条件才能使用 IndexMerge 访问：

- 对于由 `OR` 连接而成的条件，其中每个子条件都需要分别能够使用 IndexMerge 访问；
- 对于由 `AND` 连接而成的条件，其中一部分子条件需要分别能够使用 IndexMerge 访问。最终使用 IndexMerge 访问时，也只能利用这一部分条件；
- 用于构成整体 IndexMerge 计划的每个子条件，需要分别符合用于连接的 `OR`/`AND` 的语义。具体如下：
    - `json_contains` 由 `AND` 连接，符合语义；
    - `json_overlaps` 由 `OR` 连接，符合语义；
    - `json_member_of` 由 `OR`/`AND` 连接，均符合语义；
    - 包含多个值的 `json_contains` 由 `OR` 连接或包含多个值的 `json_overlaps` 由 `AND` 连接，不符合语义，而如果它们只包含一个值则符合语义。

示例如下：
```

| IndexMerge_9                  | 0.00    | root      |                                                                             | type: intersection                          |
| ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t5, index:ka(a)                                                       | range:[3,3], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t5, index:k1(cast(json_extract(`j1`, _utf8'$.path') as signed array)) | range:[1,1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t5, index:k2(cast(json_extract(`j2`, _utf8'$.path') as signed array)) | range:[2,2], keep order:false, stats:pseudo |
| └─TableRowIDScan_8(Probe)     | 0.00    | cop[tikv] | table:t5                                                                    | keep order:false, stats:pseudo              |
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
```

> 如果表达式包含多层嵌套的 `OR`/`AND`，或者表达式需要进行展开等变形之后才能直接一一对应各索引列，TiDB 可能无法使用 IndexMerge 或不能充分利用所有过滤条件，建议用户针对具体场景预先进行测试验证。以下是部分场景示例：

Reviewer 2: ```suggestion
如果表达式包含多层嵌套的 `OR`/`AND`，或者表达式需要进行展开等变形之后才能直接一一对应各索引列，TiDB 可能无法使用 IndexMerge 或不能充分利用所有过滤条件。建议针对具体场景预先进行测试验证。以下是部分场景示例：
```

```sql
CREATE TABLE t6 (a INT, j JSON, b INT, k JSON, INDEX idx(a, (CAST(j AS SIGNED ARRAY)), b), INDEX idx2(a, (CAST(k as SIGNED ARRAY)), b));
```

> 当单个 `OR` 嵌套在 `AND` 中，且需要展开表达式才能直接对应索引列的时候，TiDB 通常能够充分利用过滤条件。

Reviewer 2: ```suggestion

当单个 `OR` 嵌套在 `AND` 中，且需要展开表达式才能直接对应索引列的时候，TiDB 通常能够充分利用过滤条件。

```

| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t6, index:idx(a, cast(`j` as signed array), b)  | range:[1 1,1 1], keep order:false, stats:pseudo                                                                         |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                         |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test2.t6.a, 1), or(json_memberof(cast(1, json BINARY), test2.t6.j), json_memberof(cast(2, json BINARY), test2.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                          |
+-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
```

> 当多个 `OR` 嵌套在 `AND` 中，且需要展开表达式才能直接对应索引列的时候，TiDB 可能无法充分利用过滤条件。

Reviewer 2: ```suggestion
当多个 `OR` 嵌套在 `AND` 中，且需要展开表达式才能直接对应索引列的时候，TiDB 可能无法充分利用过滤条件。

```

6 rows in set, 1 warning (0.00 sec)
```

> 对于多个 `json_member_of`、`json_contains` 或 `json_overlaps` 由 `OR`/`AND` 连接而成的条件，需要满足以下条件才能使用 IndexMerge 访问：

Reviewer 2: ```suggestion
对于多个 `json_member_of`、`json_contains` 或 `json_overlaps` 由 `OR` 或 `AND` 连接而成的条件，需要满足以下条件才能使用 IndexMerge 访问：
```

- 对于由 `OR` 连接而成的条件，其中每个子条件都需要分别能够使用 IndexMerge 访问；
- 对于由 `AND` 连接而成的条件，其中一部分子条件需要分别能够使用 IndexMerge 访问。最终使用 IndexMerge 访问时，也只能利用这一部分条件；

> - 用于构成整体 IndexMerge 计划的每个子条件，需要分别符合用于连接的 `OR`/`AND` 的语义。具体如下：

Reviewer 2: ```suggestion
- 用于构成整体 IndexMerge 计划的每个子条件，需要分别符合用于连接的 `OR` 或 `AND` 的语义。具体如下：
```

    - `json_contains` 由 `AND` 连接，符合语义；
    - `json_overlaps` 由 `OR` 连接，符合语义；

>     - `json_member_of` 由 `OR`/`AND` 连接，均符合语义；

Reviewer 2: ```suggestion
    - `json_member_of` 由 `OR` 或 `AND` 连接，均符合语义；
```

>     - 包含多个值的 `json_contains` 由 `OR` 连接，或包含多个值的 `json_overlaps` 由 `AND` 连接，不符合语义，而如果它们只包含一个值则符合语义。

Reviewer 4: 这段有点太难懂了，下面的例子是能对应上么？

Reviewer 4: 对，我觉得一个条件拆一个例子比较好。

| IndexMerge_9                  | 0.00    | root      |                                                                             | type: intersection                          |
| ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t5, index:ka(a)                                                       | range:[3,3], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t5, index:k1(cast(json_extract(`j1`, _utf8'$.path') as signed array)) | range:[1,1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t5, index:k2(cast(json_extract(`j2`, _utf8'$.path') as signed array)) | range:[2,2], keep order:false, stats:pseudo |
| └─TableRowIDScan_8(Probe)     | 0.00    | cop[tikv] | table:t5                                                                    | keep order:false, stats:pseudo              |
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
```

> 如果表达式包含多层嵌套的 `OR`/`AND`，或者表达式需要进行展开等变形之后才能直接一一对应各索引列，TiDB 可能无法使用 IndexMerge 或不能充分利用所有过滤条件。建议针对具体场景预先进行测试验证。以下是部分场景示例：

Reviewer 2: ```suggestion
如果表达式包含多层嵌套的 `OR` 或 `AND`，或者表达式需要进行展开等变形之后才能直接一一对应各索引列，TiDB 可能无法使用 IndexMerge 或不能充分利用所有过滤条件。建议针对具体场景预先进行测试验证。以下是部分场景示例：
```

| IndexMerge_9                  | 0.20    | root      |                                                       | type: union                                                                                                             |
| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t6, index:idx(a, cast(`j` as signed array), b)  | range:[1 1,1 1], keep order:false, stats:pseudo                                                                         |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                         |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test2.t6.a, 1), or(json_memberof(cast(1, json BINARY), test2.t6.j), json_memberof(cast(2, json BINARY), test2.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                          |

> +-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+

Reviewer: 需要给个展开前的例子吗

Author: 是指啥？
现在这个 sql 就是没展开的吧，xxx and (xxx or xxx) and (xxx or xxx) 这种格式？

6 rows in set, 1 warning (0.00 sec)
```

对于多个 `json_member_of`、`json_contains` 或 `json_overlaps` 由 `OR`/`AND` 连接而成的条件，需要满足以下条件才能使用 IndexMerge 访问：

- 对于由 `OR` 连接而成的条件，其中每个子条件都需要分别能够使用 IndexMerge 访问；
- 对于由 `AND` 连接而成的条件，其中一部分子条件需要分别能够使用 IndexMerge 访问。最终使用 IndexMerge 访问时，也只能利用这一部分条件；
- 用于构成整体 IndexMerge 计划的每个子条件，需要分别符合用于连接的 `OR`/`AND` 的语义。具体如下：
    - `json_contains` 由 `AND` 连接，符合语义；
    - `json_overlaps` 由 `OR` 连接，符合语义；
    - `json_member_of` 由 `OR`/`AND` 连接，均符合语义；

>     - 包含多个值的 `json_contains` 由 `OR` 连接或包含多个值的 `json_overlaps` 由 `AND` 连接，不符合语义，而如果它们只包含一个值则符合语义。

Reviewer 2: ```suggestion
    - 包含多个值的 `json_contains` 由 `OR` 连接，或包含多个值的 `json_overlaps` 由 `AND` 连接，不符合语义，而如果它们只包含一个值则符合语义。
```
